### ISSUE TITLE
Allow copying values from the request to the response without injection

### ISSUE THREAD

**Issue by bbyars** [bbyars/mountebank#180]
Create a copyFrom behavior that can rely on jsonpath, xpath, or regex to copy values from the request and put into tokenized fields in the response

**Comment by bbyars** [comment]
Copying some of the email chain from Stephen, who originally implemented this. The initial example looked like this:

```
{
                "protocol": "http",
                "port": 2222,
                "name": "CopyFromXML",
                "requests": [],
                "stubs": [{
                                "responses": [{
                                                "is": {
                                                                "body": "<Response>\n\t<book>\n\t<title>#{PATH2}</title>\n\t<rollid>#{PATH1}</rollid>\n\t<state>#{PATH3}</state>\n\t<time>#{PATH4}</time>\n\t<publish_date>#{PATH5}</publish_date>\n\t</book>\n</Response>"
                                                },
                                                "_behaviors": {
                                                                "copyFrom": {
                                                                                "PATH1": "/book/author/text()",
                                                                                "PATH2": "/book/title/text()",
                                                                                "PATH3": "/book/genre/text()",
                                                                                "PATH4": "/book/price/text()",
                                                                                "PATH5": "/book/publish_date/text()"
                                                                }
                                                }
                                }]
                }],
                "_links": {
                                "self": {
                                                "href": "http://localhost:2525/imposters/2222"
                                }
                }
}
```

I suggested the following:

```
{
                "port": 8061,
                "protocol": "http",
                "name": "CopyFromJSON",
                "stubs": [{
                                "responses": [{
                                                "is": {
                                                                "body": "{\"Response\": \"test\",\r\n\t\"id\":#{PATH1},\r\n\t\"name\": \"#{PATH2}\",\r\n\t\"condition\": #{PATH3}}"
                                                },
                                                "_behaviors": {
                                                                "copy": {
                                                                           "from": "body",
                                                                           "to": "body",
                                                                           "values": {
                                                                                 "PATH1": { "jsonpath": "$.age" },
                                                                                "PATH2": { "jsonpath": "$.lastName" },
                                                                                "PATH3": { "jsonpath": "$.condition" }
                                                                           }
                                                                }
                                                }
                                }]
                }]
}
```



The change is more complicated syntax, but allows a few things:
* Grabbing a value out of the url (e.g. { "from": "path" }), query or headers ({ "from": "headers.accept"}, or maybe an object syntax)
* Returning the value in a header
* Working with TCP with no changes
* Adding "regex" as a selection type in addition to jsonpath and xpath

Stephen responded with:

**Comment by bbyars** [comment]
One more email:

```
{
	"copy": [{
		"from": "body",
		"jsonpath": {
			"selector": "$.age"
		},
		"into": "#{PATH1}"
	}, {
		"from": {
			"query": "param"
		},
		"xpath": {
			"selector": "$.lastName",
			"ns": {
				"isbn": "http://schemas.isbn.org/ns/1999/basic.dtd"
			}
		},
		"into": "#{PATH2}"
	}, {
		"from": {
			"headers": "X-header"
		},
		"jsonpath": {
			"selector": "$.condition"
		},
		"into": "#{PATH3}"
	}]
}
```

That does a couple things. It allows nested object keys, like query and headers, in a way that is both protocol agnostic and mirrors the ways it's done with predicates (rather than do "from": "query.param", which would introduce new syntax). It also mirrors the predicate syntax for jsonpath and xpath.

**Comment by bbyars** [comment]
Closed in f66d616
